src/expr/node_manager.cpp

```diff
@@ -97,14 +95,20 @@ typedef expr::Attribute<attr::LambdaBoundVarListTag, Node> LambdaBoundVarListAtt
 NodeManager::NodeManager()
     : d_skManager(new SkolemManager),
       d_bvManager(new BoundVarManager),
+      d_initialized(false),
       next_id(0),
       d_attrManager(new expr::attr::AttributeManager()),
       d_nodeUnderDeletion(nullptr),
       d_inReclaimZombies(false),
       d_abstractValueCount(0),
       d_skolemCounter(0)
 {
-  init();
+}
+
+NodeManager* NodeManager::currentNM()
+{
+  thread_local static NodeManager nm;
```

Comment:
Just a minor question since I'm not 100% sure how this thread_local stuff works: the change of moving this to the cpp should have no difference from before?

Reply:
That's exactly what we want, each thread will have its own instance of `nm`. Whether this is in the header or the cpp file doesn't make a difference.

Reply:
Sure, I was just wondering the motivation for the move.  This is better style I assume?

Reply:
Yep, the motivation for moving it to the source file was purely a style decision and doesn't have any effect on how `thread_local` works.